railgun/internal/ctrlutils/ingress-status.go

```diff
@@ -0,0 +1,388 @@
+package ctrlutils
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/kong/deck/file"
+	"github.com/kong/kubernetes-ingress-controller/pkg/sendconfig"
+	kicclientset "github.com/kong/kubernetes-ingress-controller/railgun/pkg/clientset"
+	"github.com/prometheus/common/log"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	knative "knative.dev/networking/pkg/apis/networking/v1alpha1"
+	knativeversioned "knative.dev/networking/pkg/client/clientset/versioned"
+	knativeApis "knative.dev/pkg/apis"
+	"knative.dev/pkg/network"
+)
+
+// dedicated function that process ingress/customer resource status update after configuration is updated within kong.
+func PullConfigUpdate(kongConfig sendconfig.Kong, log logr.Logger, ctx context.Context, kubeConfig *rest.Config, stopCh <-chan struct{}) {
```

Comment:
Rather than use a stop channel here, we should provide a context to this function and use that as the stop channel. This would be more consistent with other areas of the code, and reduce the need for managing extra controls for asynchronous processes running alongside the controller manager when we only want to bind this to the topical manager `context.Context` anyhow. Let me know your thoughts?

Reply:
StopCh is used to handle signal, for example run the KIC locally and terminate using ctrl+c. Its binding with the manager, and not the context.

Reply:
Sorry, I don't consider this resolved yet. Can you please help me to better understand why using a context would _not_ be a preferable approach here?

Reply:
Here's my suggestion for a patch that would resolve this, let me know your thoughts?
```
diff --git a/railgun/internal/ctrlutils/ingress-status.go b/railgun/internal/ctrlutils/ingress-status.go
index dd8760af..5d037c34 100644
--- a/railgun/internal/ctrlutils/ingress-status.go
+++ b/railgun/internal/ctrlutils/ingress-status.go
@@ -24,7 +24,7 @@ import (
 )
 
 // dedicated function that process ingress/customer resource status update after configuration is updated within kong.
-func PullConfigUpdate(kongConfig sendconfig.Kong, log logr.Logger, ctx context.Context, kubeConfig *rest.Config, stopCh <-chan struct{}) {
+func PullConfigUpdate(ctx context.Context, kongConfig sendconfig.Kong, log logr.Logger, kubeConfig *rest.Config) {
 	log.Info("Launching Ingress Status Update Thread.")
 	var wg sync.WaitGroup
 	for {
@@ -33,8 +33,12 @@ func PullConfigUpdate(kongConfig sendconfig.Kong, log logr.Logger, ctx context.C
 			log.Info("receive configuration information. Update ingress status %v \n", updateDone)
 			wg.Add(1)
 			go UpdateIngress(&updateDone, log, ctx, kubeConfig, &wg)
-		case <-stopCh:
-			log.Info("stop status update channel.")
+		case <-ctx.Done():
+			if err := ctx.Err(); err != nil {
+				log.Error(err, "status update thread stopped")
+			} else {
+				log.Info("status update thread stopped: context completed")
+			}
 			return
 		}
 	}
diff --git a/railgun/manager/run.go b/railgun/manager/run.go
index d387fc39..95294347 100644
--- a/railgun/manager/run.go
+++ b/railgun/manager/run.go
@@ -167,8 +167,7 @@ func Run(ctx context.Context, c *config.Config) error {
 		setupLog.Error(err, "unable to start proxy cache server")
 		return err
 	}
-	stopCh := make(chan struct{})
-	go ctrlutils.PullConfigUpdate(kongConfig, logger, ctx, kubeconfig, stopCh)
+	go ctrlutils.PullConfigUpdate(ctx, kongConfig, logger, kubeconfig)
 
 	alwaysEnabled := util.EnablementStatusEnabled
 	controllers := []ControllerDef{
```

Reply:
Yes, agree using a context.